docs/05-rfcs/2023-01-21-wing-plugins.md

```diff
@@ -0,0 +1,115 @@
+---
+title: Wing Plugins
+description: How to implement wing plugins
+---
+
+# Wing Plugin System
+
+- **Author(s)**: @Hasanaburayyan
+- **Submission Date**: {2023-01-21}
+- **Stage**: Draft
+
+Implementing a simple wing plugin system
+
+## Background
+
+Developers using wing should whenever possible only care about functional aspects of an application's infrastructure. However, there must be a way to define non-functional requirements as that will be crucial for wide adoption. For instance, an organization may have specific security requirements to adhere to for compliance and regulation sake, such as encryption at rest, versioning, replication etc... 
+
+Applying these non-functional requirements will mean that the wing compiler should provide the ability for users to write plugins that can hook into the compilation process and apply business logic to resources.
+
+## Proposal
+
+Initially the plugin system can be crafted in simple manner, allowing the ability for users to pass in additional `--plugins` option to the wing cli. These plugin files will be Javascript that contains steps for applying non-functional infrastructure requirements.
+
+A wing plugin class consists of the following methods:
+```ts
```

Comment:
I'd just go with a simple set of exported JavaScript functions instead of requiring class extension:

```js
// runs after the app is initialized, just before "synth" is called.
exports.preSynthesis = function(app) {
  // "app" is the construct tree
};

// runs after "synth" is called with the resulting terraform config
// returns an updated terraform config.
exports.postSynthesis = function(config) {
  return config; // <-- config can be mutated (optionally)
};
```

We also talked today about adding:

```js
// these hooks run after postSynthesis and 
exports.validate = function(config) {
  // does not return a value, config is immutable at this stage
}
```

We can consider (P2) first-class support for visitor pattern:

```js
exports.forEach = function(construct) {
  // here
};
```

Reply:
I need to look into this more, I have questions about how this would work using the node vm. For now I left the examples using classes but Ill get them converted once I figure this out.

Reply:
OK, I'll stay tuned :-) I don't think there should be a problem...

Reply:
So I do have a lack of experience with the Node VM library but essentially when loading in the plugin
```typescript
const script = new vm.Script(some_path);
const result = script.runInContext(some_context);
```
the result from running the script is the last line evaluated in the script. 

Thus if I had a plugin file like
```js
exports.preSynthesis = function(app) {
	// do something
};

exports.postSynthesis = function(config) {
  return config;
};
``` 
Then I would later on be able call the postSynthesis because its export was read as the last line in the script
```typescript
vm.runInNewContext('postSynthesis(config);', {
	config,
	postSynthesis: result,
});
```

However trying the same for preSynthesis would fail since the export for that was not stored in the result
```typescript
// Wont work
vm.runInNewContext('preSynthesis(config);', {
	config,
	preSynthesis: result,
});
```
(this is all just a thought dump sorry)
For me this is why I originally went with wrapping these methods in a class and exporting the class.

As well for reference I have taken much inspiration from Rocket.chat
https://www.youtube.com/watch?v=yuEKbs934e0

Reply:
hmm, yea I think thats a fair point. This being a compile time system we dont have the same security concerns for running everything in VM with limited context. 

This is maybe as you said something we can punt, thanks for the input Ill give it a whirl.

Reply:
The value I see in running in a VM is that it will force us to be more explicit about the surface area we expose to plugins and that's generally a Good Thing.

Looking at your `vm` code above, did you include an `exports` object in the VM context?

Something like:

```js
const hooks = {};
vm.runInNewContext(readFileSync(resolvedPluginModule), { exports: hooks });
hooks.preSynthesis()
```

Reply:
I had not, that does solve things 😅.